### DISCUSSION TITLE
Any way to use currently selected sdk in project that requires another version of it

### DISCUSSION THREAD

**Discussion by ZuBB** [jdx/mise#1935] [General]
Disclaimer: I am not sure if this is a known issue or a feature request, or a bug, so general Q it is

I mostly used Nodejs, so my experience is almost entirely related to it. Let's say my global version is v20, and in some projects via `.node-version`, v18 is recommended/required. So when I try to run `npm run build` (for example), I might see some complaints/warns, but at least the command will be invoked. And there is a big change that all will be totally OK. This is true for `nvm` and `fnm` managers that I used for a long time. With "mise", this is no longer a case. Even if there are no packages in the dependency tree with binding to the native package, this will not work. mise just wants to install the required version; no other options are possible.

Any thoughts on this?

**Comment by jdx** [comment]
I'm not following. Can you maybe show some example workflows of the expected and current behavior to make it clearer?

**Reply by ZuBB** [reply]
```
vv@my-precious ~/work/career-partner $ mise ls
Plugin  Version  Config Source              Requested
node    20.12.2  ~/.config/mise/config.toml 20       
vv@my-precious ~/work/career-partner $ mise current
node 20.12.2
vv@my-precious ~/work/career-partner $ cd source-code/
mise missing: node@18.18.2
vv@my-precious ~/work/career-partner/source-code $ cat .node-version
v18.18.2
vv@my-precious ~/work/career-partner/source-code $ mise ls
Plugin  Version            Config Source                                   Requested
node    18.18.2 (missing)  ~/work/career-partner/source-code/.node-version 18.18.2  
node    20.12.2                                                                     
vv@my-precious ~/work/career-partner/source-code $ mise current
mise node@18.18.2 is specified in ~/work/career-partner/source-code/.node-version, but not installed
node 18.18.2
vv@my-precious ~/work/career-partner/source-code $ npm start
mise node@18.18.2 downloading node-v18.18.2-darwin-arm64.tar.gz                          12.25 MiB/38.67 MiB (5s) ██████░░░░░░░░░░░░░░  2s
^Cfish: Unknown command: npm
vv@my-precious ~/work/career-partner/source-code $ 
```

with fnm/nvm/vfox I am still able to use 20.12.2 in dir where 18.18.2 is required

**Reply by ZuBB** [reply]
In my opinion, any tool should not do anything on its own. Unless I explicitly ask it

**Reply by jdx** [reply]
It sounds like what you're saying is that global versions should be used in place of local ones if they are not installed. If that's the case I completely disagree for the reason you just stated, "the tool should not do anything on its own unless you explicitly ask". In this case, the version being requested isn't there so mise doesn't do anything. I think it would be a mistake to assume the global version would be acceptable.

**Reply by ZuBB** [reply]
I would like to see the next behaviour (and it is consistent to nvm/fnm/vfox): currently active version should not be changed to any other automatically unless I explicitly requested it

**Reply by ZuBB** [reply]
> In this case, the version being requested 

I am sorry, but when does the presence of `.node-version` become equivalent to the word "requested"?

> so mise doesn't do anything

Its clearly not the case. Because while v20 is active (which means availability of npm CLI), running "npm start" causes the downloading of v18 (instead of running npm from v20 crowd). That can hardly be named as "anything".

**Reply by ZuBB** [reply]
> ... it would be a mistake to assume the global version would be acceptable.

Acceptable for what? 

* acceptable for an attempt to run the command -- definitely yes 
* acceptable for a successful run of the command -- maybe yes, maybe no. It depends on a few things
* acceptable for a run of the command in the proper environment -- no.

----

We just have different opinions on what a tool should do. I prefer the "do things that have been requested only" flow, while you seem to prefer an approach where the tool goes one step ahead of what the user might want to do next.

It looks like I should continue using my old tool.

**Reply by jdx** [reply]
I think what you're looking for is just disabling legacy version file parsing.

https://mise.jdx.dev/configuration.html#mise-legacy-version-file-disable-tools-node-python

**Reply by ZuBB** [reply]
Unfortunately, I am not. NVM's first release was on February 25, 2011, while mise's happened on January 26, 2023. This gives us 12 years of difference. 12 years when users were using `.nvmrc`. If I follow your advice, this means I will not be able to use mise while working on any project that still uses that file

**Reply by booniepepper** [reply]
Are you saying you want to disable `.node-version` and use `.nvmrc` but not use `.mise.toml`?

**Reply by ZuBB** [reply]
I am saying that some old projects have `.node-version`/`.nvmrc` but don't have `.mise.toml`. And if I follow the advice of the author, then I will not be able to leverage the presence of those files

**Reply by booniepepper** [reply]
Also just want to say that I am really trying to follow this, but the language here is not precise enough to know how to replicate the behavior you don't want. I've been writing JS since the 90s (remember MooTools?) and seen all sorts of tooling and libraries come and go, but I still don't feel qualified to guess what you want or whether it's going to be compatible with the goals of Mise

**Reply by ZuBB** [reply]
I might be not the best explainer, but the output from terminal (at the top) can't be read in multiple ways.

**Reply by booniepepper** [reply]
Wait what are we talking about? I thought this was a discussion about controlling versions of nodejs?

**Reply by ZuBB** [reply]
It's about ability to use currently/globally selected version ask in project that have nvmrc/node-version files with other version

**Comment by rakuyoMo** [comment]
I have an idea: 

when the dependency version required by the project does not exist, but there are other versions (whether global or otherwise), use interactive questioning to ask the user whether to use other versions (maybe higher than the project requirements, or maybe Below, I think it's possible to provide a list for the user to choose from).

For example, the project requires `node@18`, but the user did not install it. The user installed `node@17` and `node@20`. At this time, a list containing these two versions can be provided for users to choose which one to use.

As for possible subsequent command execution errors, I think that should be a problem that the user needs to solve.

Further, it is possible to add a configuration in `.mise.toml`: disallow selection of versions other than the specified version. (This depends on what the default behavior is. You can also set the default behavior to "Do not allow selection" and then manually open a new interactive question)
